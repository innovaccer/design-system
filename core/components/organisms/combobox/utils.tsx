import * as React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  inputTriggerRef?: any,
  setHighlightFirstItem?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightLastItem?: React.Dispatch<React.SetStateAction<boolean>>,
  multiSelect?: boolean,
  listRef?: any
) => {
  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      navigateOptions('up', focusedOption, setFocusedOption, listRef);
      break;
    case 'ArrowDown':
      event.preventDefault();
      navigateOptions('down', focusedOption, setFocusedOption, listRef);
      break;
    case 'Enter':
      handleEnterKey(focusedOption, multiSelect, inputTriggerRef, listRef, setFocusedOption);
      setHighlightLastItem?.(false);
      setHighlightFirstItem?.(false);
      break;
    case 'Escape':
      setOpenPopover?.(false);
      inputTriggerRef.current.focus();
      setFocusedOption?.(undefined);
      break;
    default:
      break;
  }
};

const handleEnterKey = (
  focusedOption: Element | undefined,
  multiSelect?: boolean,
  inputTriggerRef?: any,
  listRef?: any,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>
) => {
  (focusedOption as HTMLElement)?.click();
  if (!multiSelect) {
    inputTriggerRef.current.focus();
  } else {
    // to focus first option by default when last option is selected
    const listItems = listRef.current.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
    const listArr = Array.from(listItems);
    const index = listArr.findIndex((item) => {
      return item == focusedOption;
    });

    if (index === listArr.length - 1) {
      (listItems[0] as HTMLElement).focus();
      setFocusedOption && setFocusedOption(listItems[0]);
      listItems[0].scrollIntoView({ block: 'center' });
    }
  }
};

const navigateOptions = (
  direction: string,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  listRef?: any
) => {
  const listItems = listRef.current.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let index = Array.from(listItems).findIndex((item) => {
    return item == focusedOption;
  });

  if (index === -1) {
    index = direction === 'up' ? listItems.length - 1 : 0;
  } else {
    index = direction === 'up' ? (index - 1 + listItems.length) % listItems.length : (index + 1) % listItems.length;
  }

  const targetOption = listItems[index];

  (targetOption as HTMLElement).focus();
  setFocusedOption && setFocusedOption(targetOption);
  targetOption?.scrollIntoView?.({ block: 'center' });
};
