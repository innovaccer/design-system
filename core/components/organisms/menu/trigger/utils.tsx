import * as React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean | undefined>>,
  setHighlightFirstItem?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightLastItem?: React.Dispatch<React.SetStateAction<boolean>>
) => {
  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      setOpenPopover?.(true);
      setHighlightLastItem?.(true);
      break;
    case 'ArrowDown':
      event.preventDefault();
      setOpenPopover?.(true);
      setHighlightFirstItem?.(true);
      break;
    case 'Escape':
    case 'Tab':
      setOpenPopover?.(false);
      break;
    default:
      break;
  }
};

export const focusListItem = (
  position: string,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  listRef?: any
) => {
  const listItems = listRef.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let targetOption;

  if (position === 'down') {
    targetOption = listItems?.[0];
  } else {
    targetOption = listItems?.[listItems.length - 1];
  }
  (targetOption as HTMLElement)?.focus();

  if (targetOption && typeof targetOption.scrollIntoView === 'function') {
    (targetOption as HTMLElement)?.scrollIntoView({ block: 'end' });
  }
  setFocusedOption && setFocusedOption(targetOption);
};
