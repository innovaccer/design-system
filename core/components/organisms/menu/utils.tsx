import React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean | undefined>>,
  menuTriggerRef?: React.RefObject<HTMLButtonElement | null>,
  listRef?: React.RefObject<HTMLDivElement | null>,
  subListRef?: React.RefObject<HTMLDivElement | null> | null,
  isSubMenuTrigger?: boolean,
  triggerRef?: React.RefObject<HTMLDivElement | null> | React.MutableRefObject<HTMLDivElement | null>,
  menuID?: string,
  triggerID?: string,
  parentListRef?: React.RefObject<HTMLDivElement | null> | null
) => {
  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      navigateOptions('up', focusedOption, setFocusedOption, listRef);
      break;
    case 'ArrowDown':
      event.preventDefault();
      navigateOptions('down', focusedOption, setFocusedOption, listRef);
      break;
    case 'Enter':
      (focusedOption as HTMLElement)?.click();
      setOpenPopover?.(false);
      break;
    case 'Escape':
      setOpenPopover?.(false);
      if (triggerRef && !isSubMenuTrigger) {
        triggerRef?.current?.focus();
      } else {
        menuTriggerRef?.current?.focus();
      }
      setFocusedOption?.(undefined);
      break;
    case 'Tab':
      setOpenPopover?.(false);
      break;
    case 'ArrowRight':
      navigateSubMenu(isSubMenuTrigger, 'right', subListRef, menuID, triggerID, parentListRef);
      break;
    case 'ArrowLeft':
      navigateSubMenu(isSubMenuTrigger, 'left', subListRef, menuID, triggerID, parentListRef);
      break;
    default:
      break;
  }
};

const navigateOptions = (
  direction: string,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  listRef?: any
) => {
  const listItems = listRef.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let index = Array.from(listItems).findIndex((item) => {
    return item == focusedOption;
  });

  if (index === -1) {
    index = direction === 'up' ? listItems.length - 1 : 0;
  } else {
    index = direction === 'up' ? (index - 1 + listItems.length) % listItems.length : (index + 1) % listItems.length;
  }

  const targetOption = listItems[index];
  (targetOption as HTMLElement).focus();
  setFocusedOption && setFocusedOption(targetOption);
  targetOption?.scrollIntoView?.({ block: 'center' });
};

const navigateSubMenu = (
  isSubMenuTrigger?: boolean,
  direction?: string,
  subListRef?: React.RefObject<HTMLDivElement | null> | null,
  menuID?: string,
  triggerID?: string,
  parentListRef?: React.RefObject<HTMLDivElement | null> | null
) => {
  const element = document.querySelector(`[data-name="${menuID}"]`);
  const menuPlacement = element?.getAttribute('data-placement');

  if (isSubMenuTrigger) {
    if (
      (direction === 'right' && menuPlacement?.includes('right')) ||
      (direction === 'left' && menuPlacement?.includes('left'))
    ) {
      const listItems = subListRef?.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
      (listItems?.[0] as HTMLElement).focus();
    }
  } else if (
    (direction === 'left' && menuPlacement?.includes('right')) ||
    (direction === 'right' && menuPlacement?.includes('left'))
  ) {
    const triggerElement = parentListRef?.current?.querySelector(`#${triggerID}`)?.firstChild;
    (triggerElement as HTMLElement)?.focus();
  }
};
