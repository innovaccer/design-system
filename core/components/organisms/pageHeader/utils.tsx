import * as React from 'react';
import classNames from 'classnames';
import { navigationPositionType } from './PageHeader';
import { Heading, Column } from '@/index';
import styles from '@css/components/pageHeader.module.css';

export const Status = (props: {
  status: React.ReactNode;
  meta: React.ReactNode;
  navigationPosition: navigationPositionType;
  navigation: React.ReactNode;
  tabs: React.ReactNode;
}) => {
  const { status, meta, navigationPosition, navigation, tabs } = props;

  const statusClasses = classNames({
    [styles['PageHeader-statusWrapper']]: true,
    'mb-3': (navigationPosition === 'bottom' && navigation) || tabs,
  });

  return (
    <>
      {(status || meta) && (
        <div className={statusClasses} data-test="DesignSystem-PageHeader--Status">
          {status}
          {meta}
        </div>
      )}
    </>
  );
};

export const Action = (props: { actions: React.ReactNode; navigation: React.ReactNode; stepper: React.ReactNode }) => {
  const { actions, navigation, stepper } = props;

  return (
    <>
      {actions ? (
        <Column size="4" sizeXL="4" sizeM="4" data-test="DesignSystem-PageHeader--Actions">
          <div className={styles['PageHeader-actionsWrapper']}>{actions}</div>
        </Column>
      ) : (
        (navigation || stepper) && (
          <Column size="4" sizeXL="4" sizeM="4" data-test="DesignSystem-PageHeader--Actions">
            <div className={styles['PageHeader-actionsWrapper']}></div>
          </Column>
        )
      )}
    </>
  );
};

export const Nav = (props: { navigation: React.ReactNode; stepper: React.ReactNode }) => {
  const { navigation, stepper } = props;

  if (!navigation && !stepper) {
    return null;
  }
  return (
    <div className={styles['PageHeader-navigationWrapper']} data-test="DesignSystem-PageHeader--Nav">
      {navigation || stepper}
    </div>
  );
};

export const CenterNav = (props: {
  colSize: string;
  breadcrumbs: React.ReactNode;
  navigationPosition: navigationPositionType;
  navigation: React.ReactNode;
  stepper: React.ReactNode;
}) => {
  const { colSize, breadcrumbs, navigationPosition } = props;
  return (
    <>
      {(!breadcrumbs || navigationPosition === 'center') && colSize === '4' && (
        <Column size="4" sizeXL="4" sizeM="4" data-test="DesignSystem-PageHeader--CenterNav">
          <Nav {...props} />
        </Column>
      )}
    </>
  );
};

export const BackButton = (props: { button: React.ReactNode }) => {
  const { button } = props;
  return (
    <>
      {button && (
        <div className="mr-5 my-3" data-test="DesignSystem-PageHeader--Button">
          {button}
        </div>
      )}
    </>
  );
};

export const Title = (props: { badge: React.ReactNode; title: string }) => {
  const { badge, title } = props;
  return (
    <div className={styles['PageHeader-titleWrapper']} data-test="DesignSystem-PageHeader--Title">
      <Heading className={styles['PageHeader-title']}>{title}</Heading>
      {badge}
    </div>
  );
};
