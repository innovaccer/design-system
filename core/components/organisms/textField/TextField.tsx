import * as React from 'react';
import TextFieldWithTextarea, { TextFieldTextareaProps } from './TextFieldWithTextarea';
import TextFieldWithInput, { TextFieldInputProps } from './TextFieldWithInput';

export type TextFieldProps = TextFieldTextareaProps & TextFieldInputProps;

export const TextField = (props: TextFieldProps) => {
  return props.withTextarea ? <TextFieldWithTextarea {...props} /> : <TextFieldWithInput {...props} />;
};

TextField.displayName = 'TextField';
export default TextField;
