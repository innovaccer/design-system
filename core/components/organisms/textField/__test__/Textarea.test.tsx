import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { testHelper, filterUndefined, valueHelper, testMessageHelper } from '@/utils/testHelper';
import TextField, { TextFieldProps as Props } from '../TextField';

declare global {
  interface Window {
    ResizeObserver: unknown;
  }
}

const error = [true, false];
const withTextarea = [true, false];
const FunctionValue = jest.fn();

const ResizeObserver = jest.fn().mockImplementation(() => ({
  observe: jest.fn(),
  unobserve: jest.fn(),
  disconnect: jest.fn(),
}));

window.ResizeObserver = ResizeObserver;

describe('textfield component', () => {
  const mapper: Record<string, any> = {
    label: valueHelper('description', { required: true }),
    name: valueHelper('name', { required: true }),
    max: valueHelper(50, { required: true }),
    helpText: valueHelper('i am the helptext', { required: true }),
    withTextarea: valueHelper(withTextarea, { required: true, iterate: true }),
    onChange: valueHelper(FunctionValue, { required: true }),
  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as Props;

    it(testMessageHelper(attr), () => {
      const { baseElement } = render(<TextField {...attr} />);
      expect(baseElement).toMatchSnapshot();
    });
  };

  testHelper(mapper, testFunc);
});

describe('textfield component', () => {
  const mapper: Record<string, any> = {
    label: valueHelper('description', { required: true }),
    name: valueHelper('name', { required: true }),
    max: valueHelper(50, { required: true }),
    row: valueHelper(4, { required: true }),
    helpText: valueHelper('i am the helptext', { required: true }),
    withTextarea: valueHelper(true, { required: true }),
    onChange: valueHelper(FunctionValue, { required: true }),
  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as Props;

    it(testMessageHelper(attr), () => {
      const { baseElement } = render(<TextField {...attr} />);
      expect(baseElement).toMatchSnapshot();
    });
  };

  testHelper(mapper, testFunc);
});

describe('textfield component', () => {
  const mapper: Record<string, any> = {
    label: valueHelper('description', { required: true }),
    name: valueHelper('name', { required: true }),
    max: valueHelper(50, { required: true }),
    placeholder: valueHelper('helptext', { required: true }),
    error: valueHelper(error, { required: true, iterate: true }),
    helpText: valueHelper('i am the helptext', { required: true }),
    withTextarea: valueHelper(true, { required: true }),
    onChange: valueHelper(FunctionValue, { required: true }),
  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as Props;

    it(testMessageHelper(attr), () => {
      const { baseElement } = render(<TextField {...attr} />);
      expect(baseElement).toMatchSnapshot();
    });
  };

  testHelper(mapper, testFunc);
});

describe('textfield component', () => {
  const mapper: Record<string, any> = {
    label: valueHelper('description', { required: true }),
    name: valueHelper('name', { required: true }),
    defaultValue: valueHelper('default value', { required: true }),
    disabled: valueHelper([true, false], { required: true, iterate: true }),
    helpText: valueHelper('i am the helptext', { required: true }),
    withTextarea: valueHelper(true, { required: true }),
    onChange: valueHelper(FunctionValue, { required: true }),
  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as Props;

    it(testMessageHelper(attr), () => {
      const { baseElement } = render(<TextField {...attr} />);
      expect(baseElement).toMatchSnapshot();
    });
  };

  testHelper(mapper, testFunc);
});

describe('TextField component', () => {
  test('matches snapshot when input exceeds maximum character count', () => {
    const onChangeMock = jest.fn();

    const { container } = render(
      <TextField
        label="Test Label"
        name="test-name"
        placeholder="Test Placeholder"
        value="Test Value"
        defaultValue="Default Value"
        max={298}
        onChange={onChangeMock}
      />
    );
    const textarea = screen.getByPlaceholderText('Test Placeholder') as HTMLTextAreaElement;

    // Type more than the allowed character count
    fireEvent.change(textarea, { target: { value: 'a'.repeat(300) } });
    expect(container).toMatchSnapshot();
  });
});

describe('TextField Component', () => {
  test('renders with basic props', () => {
    const onChangeMock = jest.fn();
    const onBlurMock = jest.fn();
    const onFocusMock = jest.fn();
    const onClearMock = jest.fn();

    render(
      <TextField
        label="Test Label"
        name="test-name"
        placeholder="Test Placeholder"
        value="Test Value"
        defaultValue="Default Value"
        onChange={onChangeMock}
        onBlur={onBlurMock}
        onFocus={onFocusMock}
        onClear={onClearMock}
      />
    );

    // Check if the label is rendered
    expect(screen.getByText('Test Label')).toBeInTheDocument();

    // Check if the input has the correct props
    const input = screen.getByPlaceholderText('Test Placeholder') as HTMLInputElement;
    expect(input).toHaveAttribute('name', 'test-name');
    expect(input).toHaveValue('Test Value');

    // Trigger events and check if the callbacks are called
    fireEvent.change(input, { target: { value: 'New Value' } });
    expect(onChangeMock).toHaveBeenCalled();
    fireEvent.blur(input);
    expect(onBlurMock).toHaveBeenCalled();
    fireEvent.focus(input);
    expect(onFocusMock).toHaveBeenCalled();
  });

  test('renders with textarea', () => {
    render(<TextField withTextarea={true} placeholder="here is the placeholder" />);
    // Check if the component renders with textarea
    const textareaElement = screen.getByPlaceholderText('here is the placeholder');

    // Assert that the element is an exact <textarea> instance
    expect(textareaElement).toBeInTheDocument();
    expect(textareaElement).toBeInstanceOf(HTMLTextAreaElement);
    expect(textareaElement.tagName.toLowerCase()).toBe('textarea');
  });

  test('renders with help text and character count', () => {
    render(<TextField helpText="Helpful text" max={200} />);
    // Check if the help text is rendered
    expect(screen.getByText('Helpful text')).toBeInTheDocument();

    // Check if character count is initially 0/200
    expect(screen.getByText('200')).toBeInTheDocument();
  });

  test('input exceeds maximum character count', () => {
    const onChangeMock = jest.fn();
    const { getByTestId } = render(
      <TextField
        label="Test Label"
        name="test-name"
        placeholder="Test Placeholder"
        value="Test Value"
        error={true}
        defaultValue="Default Value"
        max={298}
        helpText="here is the helptext"
        onChange={onChangeMock}
      />
    );
    const textarea = screen.getByPlaceholderText('Test Placeholder') as HTMLTextAreaElement;

    // Type more than the allowed character count
    fireEvent.change(textarea, { target: { value: 'a'.repeat(300) } });
    expect(screen.getByText('300')).toBeInTheDocument();

    const searchError = getByTestId('DesignSystem-InlineMessage--Icon');
    expect(searchError).toBeInTheDocument();
  });

  test('TextField has the "input" classNames', () => {
    const onChangeMock = jest.fn();
    render(
      <TextField
        label="Test Label"
        name="test-name"
        placeholder="Test Placeholder"
        value="Test Value"
        error={true}
        defaultValue="Default Value"
        max={298}
        helpText="here is the helptext"
        onChange={onChangeMock}
      />
    );

    const textarea = screen.getByPlaceholderText('Test Placeholder') as HTMLTextAreaElement;

    expect(textarea).toHaveClass('Input-input Input-input--regular');
  });

  test('dafault value passed exceeds maximum character count with input textfield', () => {
    const onChangeMock = jest.fn();
    const { getByTestId } = render(
      <TextField
        label="Test Label"
        name="test-name"
        placeholder="Test Placeholder"
        value="Test Value"
        error={true}
        defaultValue="Default Value"
        max={5}
        helpText="here is the helptext"
        onChange={onChangeMock}
      />
    );

    const searchError = getByTestId('DesignSystem-InlineMessage--Icon');
    expect(searchError).toBeInTheDocument();
  });
  test('dafault value passed exceeds maximum character count with textarea textfield', () => {
    const onChangeMock = jest.fn();
    const { getByTestId } = render(
      <TextField
        label="Test Label"
        name="test-name"
        placeholder="Test Placeholder"
        value="Test Value"
        error={true}
        withTextarea={true}
        defaultValue="Default Value"
        max={5}
        helpText="here is the helptext"
        onChange={onChangeMock}
      />
    );

    const searchError = getByTestId('DesignSystem-InlineMessage--Icon');
    expect(searchError).toBeInTheDocument();
  });
});
