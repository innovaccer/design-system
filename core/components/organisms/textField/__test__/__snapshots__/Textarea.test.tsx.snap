// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField component matches snapshot when input exceeds maximum character count 1`] = `
<div>
  <div>
    <div
      class="Label Label--withInput"
      data-test="DesignSystem-Label"
    >
      <label
        class="Label-text"
        data-test="DesignSystem-Label--Text"
      >
        Test Label
      </label>
    </div>
    <div
      class="Input Input--regular Input--error"
      data-test="DesignSystem-InputWrapper"
      role="presentation"
      style="min-width: 256px;"
    >
      <input
        class="Input-input Input-input--regular"
        data-test="DesignSystem-Input"
        label="Test Label"
        max="298"
        name="test-name"
        placeholder="Test Placeholder"
        type="text"
        value="Test Value"
      />
    </div>
    <div
      class="d-flex justify-content-between"
      style="min-width: 256px;"
    >
      <div
        class="InlineMessage mt-3 d-flex"
        data-test="DesignSystem-InlineMessage"
      >
        <i
          class="material-symbols material-symbols-rounded Icon Icon--alert InlineMessage-icon--small"
          data-test="DesignSystem-InlineMessage--Icon"
          role="button"
          style="font-size: 14px; width: 14px;"
        >
          error
        </i>
        <span
          class="Text Text--default Text--medium Text--small InlineMessage-text--alert"
          data-test="DesignSystem-InlineMessage--Description"
        >
           
        </span>
      </div>
      <div
        class="mt-3 d-flex"
      >
        <span
          class="Text Text--medium Text--small color-alert pr-2"
          data-test="DesignSystem-Text"
        >
          300
        </span>
        <span
          class="Text Text--subtle Text--medium Text--small pr-2"
          data-test="DesignSystem-Text"
        >
          /
        </span>
        <span
          class="Text Text--subtle Text--medium Text--small"
          data-test="DesignSystem-Text"
        >
          298
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`textfield component 
  label: "description", name: "name", defaultValue: "default value", disabled: false, helpText: "i am the helptext", withTextarea: true, onChange: "[Function]"
 1`] = `
<body>
  <div>
    <div>
      <div
        class="Label Label--withInput"
        data-test="DesignSystem-Label"
      >
        <label
          class="Label-text"
          data-test="DesignSystem-Label--Text"
        >
          description
        </label>
      </div>
      <textarea
        class="Textarea Textarea--resize"
        data-test="DesignSystem-Textarea"
        helptext="i am the helptext"
        label="description"
        name="name"
        rows="3"
      >
        default value
      </textarea>
      <div
        class="d-flex justify-content-between"
        style="width: 0px;"
      >
        <div
          class="mt-3 d-flex"
        >
          <span
            class="Text Text--subtle Text--medium Text--small"
            data-test="DesignSystem-Text"
          >
            i am the helptext
          </span>
        </div>
        <div
          class="mt-3 d-flex"
        >
          <span
            class="Text Text--subtle Text--medium Text--small pr-2"
            data-test="DesignSystem-Text"
          >
            0
          </span>
          <span
            class="Text Text--subtle Text--medium Text--small pr-2"
            data-test="DesignSystem-Text"
          >
            /
          </span>
          <span
            class="Text Text--subtle Text--medium Text--small"
            data-test="DesignSystem-Text"
          >
            200
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`textfield component 
  label: "description", name: "name", defaultValue: "default value", disabled: true, helpText: "i am the helptext", withTextarea: true, onChange: "[Function]"
 1`] = `
<body>
  <div>
    <div>
      <div
        class="Label Label--withInput"
        data-test="DesignSystem-Label"
      >
        <label
          class="Label-text"
          data-test="DesignSystem-Label--Text"
        >
          description
        </label>
      </div>
      <textarea
        class="Textarea Textarea--resize"
        data-test="DesignSystem-Textarea"
        disabled=""
        helptext="i am the helptext"
        label="description"
        name="name"
        rows="3"
      >
        default value
      </textarea>
      <div
        class="d-flex justify-content-between"
        style="width: 0px;"
      >
        <div
          class="mt-3 d-flex"
        >
          <span
            class="Text Text--subtle Text--medium Text--small"
            data-test="DesignSystem-Text"
          >
            i am the helptext
          </span>
        </div>
        <div
          class="mt-3 d-flex"
        >
          <span
            class="Text Text--subtle Text--medium Text--small pr-2"
            data-test="DesignSystem-Text"
          >
            0
          </span>
          <span
            class="Text Text--subtle Text--medium Text--small pr-2"
            data-test="DesignSystem-Text"
          >
            /
          </span>
          <span
            class="Text Text--subtle Text--medium Text--small"
            data-test="DesignSystem-Text"
          >
            200
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`textfield component 
  label: "description", name: "name", max: 50, helpText: "i am the helptext", withTextarea: false, onChange: "[Function]"
 1`] = `
<body>
  <div>
    <div>
      <div
        class="Label Label--withInput"
        data-test="DesignSystem-Label"
      >
        <label
          class="Label-text"
          data-test="DesignSystem-Label--Text"
        >
          description
        </label>
      </div>
      <div
        class="Input Input--regular"
        data-test="DesignSystem-InputWrapper"
        role="presentation"
        style="min-width: 256px;"
      >
        <input
          class="Input-input Input-input--regular"
          data-test="DesignSystem-Input"
          helptext="i am the helptext"
          label="description"
          max="50"
          name="name"
          type="text"
        />
      </div>
      <div
        class="d-flex justify-content-between"
        style="min-width: 256px;"
      >
        <div
          class="mt-3 d-flex"
        >
          <span
            class="Text Text--subtle Text--medium Text--small"
            data-test="DesignSystem-Text"
          >
            i am the helptext
          </span>
        </div>
        <div
          class="mt-3 d-flex"
        >
          <span
            class="Text Text--subtle Text--medium Text--small pr-2"
            data-test="DesignSystem-Text"
          >
            0
          </span>
          <span
            class="Text Text--subtle Text--medium Text--small pr-2"
            data-test="DesignSystem-Text"
          >
            /
          </span>
          <span
            class="Text Text--subtle Text--medium Text--small"
            data-test="DesignSystem-Text"
          >
            50
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`textfield component 
  label: "description", name: "name", max: 50, helpText: "i am the helptext", withTextarea: true, onChange: "[Function]"
 1`] = `
<body>
  <div>
    <div>
      <div
        class="Label Label--withInput"
        data-test="DesignSystem-Label"
      >
        <label
          class="Label-text"
          data-test="DesignSystem-Label--Text"
        >
          description
        </label>
      </div>
      <textarea
        class="Textarea Textarea--resize"
        data-test="DesignSystem-Textarea"
        helptext="i am the helptext"
        label="description"
        max="50"
        name="name"
        rows="3"
      />
      <div
        class="d-flex justify-content-between"
        style="width: 0px;"
      >
        <div
          class="mt-3 d-flex"
        >
          <span
            class="Text Text--subtle Text--medium Text--small"
            data-test="DesignSystem-Text"
          >
            i am the helptext
          </span>
        </div>
        <div
          class="mt-3 d-flex"
        >
          <span
            class="Text Text--subtle Text--medium Text--small pr-2"
            data-test="DesignSystem-Text"
          >
            0
          </span>
          <span
            class="Text Text--subtle Text--medium Text--small pr-2"
            data-test="DesignSystem-Text"
          >
            /
          </span>
          <span
            class="Text Text--subtle Text--medium Text--small"
            data-test="DesignSystem-Text"
          >
            50
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`textfield component 
  label: "description", name: "name", max: 50, placeholder: "helptext", error: false, helpText: "i am the helptext", withTextarea: true, onChange: "[Function]"
 1`] = `
<body>
  <div>
    <div>
      <div
        class="Label Label--withInput"
        data-test="DesignSystem-Label"
      >
        <label
          class="Label-text"
          data-test="DesignSystem-Label--Text"
        >
          description
        </label>
      </div>
      <textarea
        class="Textarea Textarea--resize"
        data-test="DesignSystem-Textarea"
        helptext="i am the helptext"
        label="description"
        max="50"
        name="name"
        placeholder="helptext"
        rows="3"
      />
      <div
        class="d-flex justify-content-between"
        style="width: 0px;"
      >
        <div
          class="mt-3 d-flex"
        >
          <span
            class="Text Text--subtle Text--medium Text--small"
            data-test="DesignSystem-Text"
          >
            i am the helptext
          </span>
        </div>
        <div
          class="mt-3 d-flex"
        >
          <span
            class="Text Text--subtle Text--medium Text--small pr-2"
            data-test="DesignSystem-Text"
          >
            0
          </span>
          <span
            class="Text Text--subtle Text--medium Text--small pr-2"
            data-test="DesignSystem-Text"
          >
            /
          </span>
          <span
            class="Text Text--subtle Text--medium Text--small"
            data-test="DesignSystem-Text"
          >
            50
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`textfield component 
  label: "description", name: "name", max: 50, placeholder: "helptext", error: true, helpText: "i am the helptext", withTextarea: true, onChange: "[Function]"
 1`] = `
<body>
  <div>
    <div>
      <div
        class="Label Label--withInput"
        data-test="DesignSystem-Label"
      >
        <label
          class="Label-text"
          data-test="DesignSystem-Label--Text"
        >
          description
        </label>
      </div>
      <textarea
        class="Textarea Textarea--resize Textarea--error"
        data-test="DesignSystem-Textarea"
        helptext="i am the helptext"
        label="description"
        max="50"
        name="name"
        placeholder="helptext"
        rows="3"
      />
      <div
        class="d-flex justify-content-between"
        style="width: 0px;"
      >
        <div
          class="InlineMessage mt-3 d-flex"
          data-test="DesignSystem-InlineMessage"
        >
          <i
            class="material-symbols material-symbols-rounded Icon Icon--alert InlineMessage-icon--small"
            data-test="DesignSystem-InlineMessage--Icon"
            role="button"
            style="font-size: 14px; width: 14px;"
          >
            error
          </i>
          <span
            class="Text Text--default Text--medium Text--small InlineMessage-text--alert"
            data-test="DesignSystem-InlineMessage--Description"
          >
            i am the helptext
          </span>
        </div>
        <div
          class="mt-3 d-flex"
        >
          <span
            class="Text Text--subtle Text--medium Text--small pr-2"
            data-test="DesignSystem-Text"
          >
            0
          </span>
          <span
            class="Text Text--subtle Text--medium Text--small pr-2"
            data-test="DesignSystem-Text"
          >
            /
          </span>
          <span
            class="Text Text--subtle Text--medium Text--small"
            data-test="DesignSystem-Text"
          >
            50
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`textfield component 
  label: "description", name: "name", max: 50, row: 4, helpText: "i am the helptext", withTextarea: true, onChange: "[Function]"
 1`] = `
<body>
  <div>
    <div>
      <div
        class="Label Label--withInput"
        data-test="DesignSystem-Label"
      >
        <label
          class="Label-text"
          data-test="DesignSystem-Label--Text"
        >
          description
        </label>
      </div>
      <textarea
        class="Textarea Textarea--resize"
        data-test="DesignSystem-Textarea"
        helptext="i am the helptext"
        label="description"
        max="50"
        name="name"
        row="4"
        rows="3"
      />
      <div
        class="d-flex justify-content-between"
        style="width: 0px;"
      >
        <div
          class="mt-3 d-flex"
        >
          <span
            class="Text Text--subtle Text--medium Text--small"
            data-test="DesignSystem-Text"
          >
            i am the helptext
          </span>
        </div>
        <div
          class="mt-3 d-flex"
        >
          <span
            class="Text Text--subtle Text--medium Text--small pr-2"
            data-test="DesignSystem-Text"
          >
            0
          </span>
          <span
            class="Text Text--subtle Text--medium Text--small pr-2"
            data-test="DesignSystem-Text"
          >
            /
          </span>
          <span
            class="Text Text--subtle Text--medium Text--small"
            data-test="DesignSystem-Text"
          >
            50
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;
